Solution of task - Group people by city

zulqer-nain opened the pull request:
- Added solution to group people by city.
- Added solution to group people by gender.
- Added generic function "groupBy" to group by any property .
- Added test case for  group people by city
- Added test case for  group people by gender.

zulqer-nain committed:
NarcissisticNumbers solution and test case.

Solution to find Narcisstic numbers in range .
Solution to find if provided number is Narcisstic or not .
Test cases to check whether provided number is Narcisstic or not .
Test cases to find Narcisstic numbers in provided range .

zulqer-nain committed:
Resolved Comments on PR

Replaced annonymus function with arrow function .
Added negative number test case.
Added Negative Range test case.
Improvised isNarcissisticNumber function .

zulqer-nain committed:
Solution of Group People By City

Added solution to group people by city.
Added solution to group people by gender.
Added generic function "groupBy" to group by any property .
Added test case for  group people by city
Added test case for  group people by gender.

krajasekhar-unify commented on src/zulqer/groupPeople/index.ts:
```diff
@@ -0,0 +1,62 @@
+  people: Array<People>;
+}
+
+const groupBy = (arrayOfObject: Array<any>, property: string): Array<any> => {
```
- Since this function is being called with `peopleArr` as the first argument, wouldn't the type of the first argument be `Array<People>` instead of `Array<any>` ?
- Can we have a better naming for the variable `arrayOfObject`?

zulqer-nain replied:
purpose of this function is to group any kind of Array not just Array of People and can be grouped by any property/key

krajasekhar-unify replied:
Understood. Can we still restrict it as array of `objects` instead of `any`?

zulqer-nain replied:
Sure.

krajasekhar-unify commented on src/zulqer/groupPeople/index.ts:
```diff
@@ -0,0 +1,62 @@
+
+const groupBy = (arrayOfObject: Array<any>, property: string): Array<any> => {
+  const grouped = [];
+  arrayOfObject.map(function (obj) {
```
Please try to use arrow function inside `map`.
```
arrayOfObject.map(obj => { });
```

krajasekhar-unify commented on src/zulqer/groupPeople/index.ts:
```diff
@@ -0,0 +1,62 @@
+    if (grouped[obj[property] || property]) {
+      grouped[obj[property] || property].push(obj);
+    } else {
+      grouped[obj[property] || property] = [obj];
```
- `grouped` shall be an object, using an array as an object to store at different keys than numeric indices is not a good practice. 
- if the given data `obj` doesn't have the property (say `city`), there is a handling added to assign with the key as property name (`grouped[property]`). Those will be stored at a key with name as `city` which is not good. If the data `obj` doesn't have the property we shall either skip the record or store with a key like `Unassigned`

krajasekhar-unify commented on src/zulqer/narcissisticNumbers/__tests__/index.test.ts:
```diff
@@ -0,0 +1,42 @@
+import { getNarcissisticNumbersInRange, isNarcissisticNumber } from '..';
```
Since there is a different pull request exist for `narcissisticNumbers`, please remove this file from this pull request. As both are independent features there is a chance one pull request will be promoted before other.

krajasekhar-unify commented on src/zulqer/narcissisticNumbers/index.ts:
```diff
@@ -0,0 +1,22 @@
+function isNarcissisticNumber(num: number): boolean {
```
Since there is a different pull request exist for `narcissisticNumbers`, please remove this file from this pull request. As both are independent features there is a chance one pull request will be promoted before other.

krajasekhar-unify reviewed (changes requested):
Please make the requested changes.

krajasekhar-unify commented:
Also please link the original issue for this pull request in the `Linked Issues` section.

zulqer-nain committed:
Resolved comments

Resolved comments
changed to arroe function
improved return type

krajasekhar-unify commented on src/zulqer/groupPeople/__test__/index.test.ts:
```diff
@@ -0,0 +1,24 @@
+import { getPeopleByCity, getPeopleByGender } from '../index';
```
Lets have folder name `__tests__` instead of `__test__`
![image](https://user-images.githubusercontent.com/71023607/93057108-069c1200-f68b-11ea-8cb3-14af6ba2bdc7.png)

nirajunify replied:
No. Let's keep unit test cases in the same directory as code.

zulqer-nain committed:
Reolved comments

Restructure test file

krajasekhar-unify commented on src/zulqer/groupPeople/groupPeople.test.ts:
```diff
@@ -0,0 +1,24 @@
+  });
+});
+
+describe('group by gender function', function () {
```
Arrow function please.

krajasekhar-unify commented on src/zulqer/groupPeople/data.ts:
```diff
@@ -0,0 +1,159 @@
+const data = [
```
@nirajunify we have a file here to store the test data and is useful only for test purposes. Please advise on where to keep and the filename.

nirajunify replied:
Since this data is problem-specific, and the current solution requires it immediately - keep the data file in the same directory as the solution. However, datafile should be a JSON file instead of ts.

krajasekhar-unify reviewed (changes requested):
Also the commit messages shall convey **what is the fix done** in that commit instead of `Resolving for comments mentioned` as people don't know what was the comment made before.

zulqer-nain committed:
Resolved comments

- replaced function to arrow function
- removed __tests__ directory
-improved generic function "groupBy"
 specified return type
returning grouped data only if property exists

krajasekhar-unify reviewed (approved):
Approving. Naming the commits shall be better.

zulqer-nain committed:
Merge branch 'zulqer-nain-task/zulqer-city-people-grouping'
